### PR TITLE
add miniPile dataset for pretraining, 1M entries (solves the 'out of data' at 40 iters issue)

### PR DESCRIPTION
**Pull request by lessw2020**
This PR add's minipile (1M, 6GB) dataset as an option for pretraining with torchtrain. 
It resolves the issue where we run out of data after 40 iterations with the default alpaca dataset. 
Per @tianyu-l's excellent suggestion, have refactored to have a single hf_datasets.py file that supports both minipile and alpaca since it turned out no need for any different tokenizer, etc. 
Also cleaned up the datasets package so that create_tokenizer is exposed directly, and thus all public apis can be used directly from torchtrain.datasets.
Lastly - added warning if/when a dataset is being re-looped so users don't get burned by overfitting:
Screenshot 2024-03-06 at 5 11 09 AM

Adds a color highlight to showcase what dataloader was built:
Screenshot 2024-03-05 at 9 19 10 PM
and
Screenshot 2024-03-05 at 9 22 01 PM

Usage:
just add "minipile" or "alpaca" as the dataset in the training config toml file. 
Screenshot 2024-02-25 at 12 35 26 PM

Testing:
verified training loss is improving and ran for 100 iters to verify no issue with out of data any longer with minipile.
reran with alpaca and saw the expected out of data at 40 iters without infinite loop option, runs to 100 with infinite.

Notes:
I did not make this a default dataset since for debugmodel, mostly running 10 iters is fine and there's 6GB to pull down.
Screenshot 2024-02-25 at 12 30 29 PM

### CODE REVIEW

File: torchtrain/datasets/minipile.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import torch
+from torch.utils.data import DataLoader, IterableDataset
+
+from torchtrain.datasets.tokenizer import TokenizerIf
+
+from datasets import load_dataset
+from datasets.distributed import split_dataset_by_node
+
+
+class MiniPileDataset(IterableDataset):
```
Comment (tianyu-l) [inline torchtrain/datasets/minipile.py]: Since most of the code are sharable, to have better maintainability shall we abstract this and have a single IterableDataset e.g. HuggingFaceDataset or just HFDataset? We can keep two build functions and send in "JeanKaddour/minipile" as an arg. This would make adding other dataset from HF much eaiser.

File: torchtrain/datasets/__init__.py [hunk 2]
```diff
@@ -2,10 +2,15 @@
 # This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.
 
 from torchtrain.datasets.alpaca import build_alpaca_data_loader
-from torchtrain.datasets.tokenizer import create_tokenizer
+from torchtrain.datasets.minipile import build_minipile_data_loader
 
-__all__ = ["build_alpaca_data_loader", "create_tokenizer", "pad_batch_to_longest_seq"]
+__all__ = [
+    "build_alpaca_data_loader",
+    "build_minipile_data_loader" "create_tokenizer",
```
Comment (tianyu-l) [inline torchtrain/datasets/__init__.py]: missing a `,` here?

File: torchtrain/datasets/__init__.py [hunk 3]
```diff
@@ -2,10 +2,15 @@
...
+    "pad_batch_to_longest_seq",
```
Comment (wanchaol) [inline torchtrain/datasets/__init__.py]: I think we deleted this import already, so this is probably a leftover that I forgot to delete lol

File: torchtrain/datasets/minipile.py (hunk 1 above)
Comment (lessw2020) [reply torchtrain/datasets/minipile.py]: yes, will work on this single impl now.  I originally thought minipile would require more extensive differences so that's how I ended up with two files, but it's a good idea to consolidate for ease of re-use.

Comment (lessw2020) [reply torchtrain/datasets/minipile.py]: all moved to single file impl (hf_datasets.py).

File: torchtrain/datasets/__init__.py (hunk 2 above)
Comment (lessw2020) [reply torchtrain/datasets/__init__.py]: yes, fixed!

File: torchtrain/datasets/__init__.py (hunk 3 above)
Comment (lessw2020) [reply torchtrain/datasets/__init__.py]: yes correct!  I have removed.

File: torchtrain/datasets/alpaca.py [hunk 4]
```diff
@@ -1,90 +0,0 @@
...
-        self._data = split_dataset_by_node(ds, rank, world_size)
-        self._tokenizer = tokenizer
-        self.seq_len = seq_len
-        self.infinite = infinite
-
-    def __iter__(self):
-        max_buffer_token_len = 1 + self.seq_len
-        all_tokens: List[int] = []
-
-        while True:
```
Comment (tianyu-l) [inline torchtrain/datasets/alpaca.py]: It seems the added file is based on the version before infinite loop was supported. Let's update it.

File: torchtrain/datasets/hf_datasets.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    {
+        "instruction": "Create a classification task by clustering the given list of items.",
+        "input": "Apples, oranges, bananas, strawberries, pineapples",
+        "output": "Class 1: Apples, Oranges\nClass 2: Bananas, Strawberries\nClass 3: Pineapples",
+        "text": "Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nCreate a classification task by clustering the given list of items.\n\n### Input:\nApples, oranges, bananas, strawberries, pineapples\n\n### Response:\nClass 1: Apples,
+        Oranges\nClass 2: Bananas, Strawberries\nClass 3: Pineapples",  # noqa: B950
+    }
+
+    Example:
+    >>> minipile_ds = MiniPileDataset(tokenizer=tokenizer)
```
Comment (tianyu-l) [inline torchtrain/datasets/hf_datasets.py]: MiniPile -> HuggingFace

File: torchtrain/datasets/__init__.py [hunk 6]
```diff
@@ -1,11 +1,14 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates.
 # This software may be used and distributed according to the terms of the Llama 2 Community License Agreement.
 
-from torchtrain.datasets.alpaca import build_alpaca_data_loader
-from torchtrain.datasets.tokenizer import create_tokenizer
```
Comment (tianyu-l) [inline torchtrain/datasets/__init__.py:5]: Curious why we don't want `create_tokenizer` directly under `torchtrain.datasets`? Also if we remove it here, we should remove it from `__all__` as well.

File: torchtrain/datasets/hf_datasets.py (hunk 5 above)
Comment (tianyu-l) [reply torchtrain/datasets/hf_datasets.py]: let's still use alpaca in the example since mentally it feels "simpler"

File: torchtrain/datasets/alpaca.py (hunk 4 above)
Comment (lessw2020) [reply torchtrain/datasets/alpaca.py]: good catch!  restored and verified infinite working by running alpaca with 100 iters.

File: torchtrain/datasets/hf_datasets.py (hunk 5 above)
Comment (lessw2020) [reply torchtrain/datasets/hf_datasets.py]: fixed and updated.

File: torchtrain/datasets/__init__.py (hunk 6 above)
Comment (lessw2020) [reply torchtrain/datasets/__init__.py:5]: this is just an artifact from earlier but agree that it should be consistent.
I've updated the __init__.py to integrate create tokenizer there (it was missing which was why the reference in train was directly calling it).
updated train.py to only call from the datasets package...looks much cleaner now!
